cabalscan/src/CabalScan/Options.hs

```diff
@@ -5,32 +5,47 @@ module CabalScan.Options
   , parseCommandLine
   ) where
 
-import Options.Applicative
+import Data.List (intercalate)
 import Path (Abs, File, Path, parseAbsFile)
+import System.Environment
+import System.Exit
+import System.IO (hPutStrLn, stderr)
+
+type AbsoluteFilepath = Path Abs File
 
 data Options = Options
-  { cabalFiles :: [Path Abs File]
+  { cabalFiles :: [AbsoluteFilepath]
   }
 
+data InvocationError = MissingFiles | WrongFilePath String | PrintHelp
+
 parseCommandLine :: IO Options
-parseCommandLine = execParser $
-    info
-      (parser <**> helper)
-      (fullDesc
-         <> progDesc
-              (mconcat
-                 [ "Prints in stdout information extracted from Cabal files"
-                 , " in JSON format."
-                 ]
-              )
-         <> header "cabalscan - extract build information from Cabal files"
-      )
-  where
-    parser = Options
-      <$> some (argument someFile (metavar "CABAL_FILES..."))
+parseCommandLine = do
+  args <- getArgs
+  case parseArgs args [] of
+    Right xs -> return (Options {cabalFiles = xs})
+    Left err -> printMsgAndExit err
 
-    someFile :: ReadM (Path Abs File)
-    someFile = eitherReader $ \s ->
-      case parseAbsFile s of
-        Just f -> Right f
-        Nothing -> Left $ "couldn't parse absolute file path: " ++ s
+parseArgs :: [String] -> [AbsoluteFilepath] -> Either InvocationError [AbsoluteFilepath]
+parseArgs [] [] = Left MissingFiles
+parseArgs [] xs = Right xs
+parseArgs (x:xs) fs | x `elem` ["-h", "--help"] = Left PrintHelp
+                    | otherwise = case parseAbsFile x of
+                                    Just f -> parseArgs xs (f:fs)
+                                    _      -> Left $ WrongFilePath x
```

Comment:
Probably it doesn't matter a lot, but I think this function is returning the files in reversed order.

Reply:
If you use `Either String (Path Abs File)" as the return type, instead of `Maybe`, I think you can collect an error message from `parseAbsFile`. I'm not sure how helpful it is going to be.

Reply:
Will check that. Thx.

Reply:
@facundominguez 

Is it better in your opinion?
```haskell
import Data.Bifunctor (first)

...

parseArgs (x:xs) fs | x `elem` ["-h", "--help"] = Left PrintHelp
                    | otherwise = first (WrongFilePath . show) (parseAbsFile x) >>= \f -> parseArgs xs (f:fs)
```

Then it produces output like:
```
Couldn't parse absolute file path: InvalidAbsFile "asdfdsa"

Usage: cabalscan CABAL_FILES...
  Prints in stdout information extracted from Cabal files in JSON format.
```

Reply:
or...

```
bimap (\_ -> WrongFilePath x) (:fs) (parseAbsFile x) >>= parseArgs xs
```
This one is better IMHO. It's more concise and doesn't print the name of the exception.

Reply:
How about doing it without recursion
```Haskell
parseArgs :: [String] -> Either InvocationError [AbsoluteFilepath]
parseArgs [] = Left MissingFiles
parseArgs xs = sequence [ maybe (Left $ WrongFilePath x) Right (parseAbsFile x) | x <- xs ]
```
?